### PR TITLE
Avoid raw pointers usage (#1360)

### LINKED ISSUE #1360: Improve exception safety with smart pointers
Would you like to [wrap any pointer data members](https://github.com/falcosecurity/falco/blob/e46641d24d74fbe7921039834fdf5b01c4715009/userspace/engine/rules.cpp#L39 "Update candidate: falco_rules class") with the class template “[std::unique_ptr](https://en.wikipedia.org/wiki/Smart_pointer#unique_ptr "Description for the usage of smart pointers")”?

### PR DESCRIPTION
**Pull request by deepskyblue86**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area rules

> /area tests

> /area proposals

**What this PR does / why we need it**:
Adoption of unique_ptr for better readability and safety
**Which issue(s) this PR fixes**:

Fixes #1360

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: userspace/engine/falco_utils.h [hunk 1]
```diff
@@ -26,6 +26,17 @@ limitations under the License.
 
 #pragma once
 
+#if (__cplusplus < 201402L)
```
Comment (leodido) [inline userspace/engine/falco_utils.h:29]: We should already have `absl::make_unique` available if I don't remember it wrong.

This way we could avoid the polyfill here.

Would you please look into it?

**Review by leodido — COMMENTED** [review]
Thanks for this Angelo!

Left a comment :)

File: userspace/engine/falco_utils.h (hunk 1 above)
Comment (deepskyblue86) [reply userspace/engine/falco_utils.h:29]: I can't find it used anywhere. I do see it as a gRPC submodule, meaning its use would conflict MINIMAL_BUILD.
Do you prefer to use abseil and have it as a separate ExternalProject?

Comment (leodido) [reply userspace/engine/falco_utils.h:29]: Right! I was not remembering that it comes as gRPC submodule...

In this case, I'd left your polyfill.

Anyways, if we're talking about what I prefer. Well, I'd prefer to port the whole codebase to C++14 (or greater). 🤣

File: userspace/engine/ruleset.h [hunk 2]
```diff
@@ -106,18 +108,18 @@ class falco_ruleset
...
-		std::vector<std::list<filter_wrapper *> *> m_filter_by_event_tag;
+		std::vector<std::unique_ptr<filter_wrapper_list>> m_filter_by_event_tag;
 
 	};
 
-	std::vector<ruleset_filters *> m_rulesets;
+	std::vector<std::unique_ptr<ruleset_filters>> m_rulesets;
 
 	// Maps from tag to list of filters having that tag.
-	std::map<std::string, std::list<filter_wrapper *>> m_filter_by_event_tag;
```
Comment (deepskyblue86) [inline userspace/engine/ruleset.h:113]: @mstemm do you recall why we have a list here but a list pointer in `ruleset_filters::m_filter_by_event_tag`?
I don't want to change it in this PR but I'm curious...

**Pull request closed without merging**